src/repo/mod.rs

```diff
@@ -111,10 +114,67 @@ pub trait DataStore: Debug + Send + Sync + Unpin + 'static {
     async fn wipe(&self);
 }
 
+#[async_trait]
+pub trait PinStore: Debug + Send + Sync + Unpin + 'static {
+    async fn is_pinned(&self, block: &Cid) -> Result<bool, Error>;
+
+    async fn insert_pin(&self, target: &Cid, kind: PinKind<&'_ Cid>) -> Result<(), Error>;
+
+    async fn remove_pin(&self, target: &Cid, kind: PinKind<&'_ Cid>) -> Result<(), Error>;
+
+    async fn list(
+        &self,
+        mode: Option<PinMode>,
+    ) -> futures::stream::BoxStream<'static, Result<(Cid, PinMode), Error>>;
+
+    // here we should have resolved ids
+    // go-ipfs: doesnt start fetching the paths
+    // js-ipfs: starts fetching paths
+    // FIXME: there should probably be an additional Result<$inner, Error> here; the per pin error
+    // is serde OR cid::Error.
+    /// Returns error if any of the ids isn't pinned in the required type, otherwise returns
+    /// the pin details if all of the cids are pinned in one way or the another.
+    async fn query(
+        &self,
+        ids: Vec<Cid>,
+        requirement: Option<PinMode>,
+    ) -> Result<Vec<(Cid, PinKind<Cid>)>, Error>;
+}
+
 #[derive(Clone, Copy, Debug)]
 pub enum Column {
     Ipns,
-    Pin,
+}
+
+/// `PinMode` is the description of pin type for quering purposes.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum PinMode {
+    Indirect,
+    Direct,
+    Recursive,
+}
+
+impl<'a, B: Borrow<Cid>> PartialEq<&'a PinMode> for PinKind<B> {
+    fn eq(&self, other: &&PinMode) -> bool {
```

Comment:
just making sure that the double reference is needed here, especially since the impl is already for `&'a PinMode`

Reply:
yeah was wondering this myself as well. I did try without first of course. Could try again or at least put a FIXME.

Reply:
can be removed if I derefence at callsite.

TBH I have no idea what's going on here, I can still dereference the `PinMode` if I remove the `Copy` bound (which I will just to be overly conservative and it's not really needed). How can I deref without copying or cloning for equality comparison...?

I guess the only plausible explanation is that the my inserted deref is just for trait resolving, but the actual comparison will happen by again borrowing the deref `&*PinMode` to produce a match for the `fn eq(&self, other: &PinMode)` and following that there is a "simplify" phase which gets rid of extra deref-borrow pairs when they have no side effects?